Python/fileutils.c

```diff
@@ -922,6 +922,86 @@ _Py_GetLocaleEncodingObject(void)
     return str;
 }
 
+#ifdef HAVE_NON_UNICODE_WCHAR_T_REPRESENTATION
+
+/* Convert a wide character string to the UTF32 encoded char32_t string. This
+   is necessary on systems where internal form of wchar_t is not already
+   Unicode (e.g. Oracle Solaris).
+
+   Return a pointer to a newly allocated char32_t string, use PyMem_Free() to
+   free the memory. Return NULL and raise exception on conversion or memory
+   allocation error. */
+char32_t*
+_Py_convert_wchar_t_to_UTF32(const wchar_t* u, Py_ssize_t size)
+{
+    /* Ensure we won't overflow the size. */
+    if (size > ((PY_SSIZE_T_MAX / (Py_ssize_t)sizeof(wchar_t)) - 1)) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    /* Given 'u' might not be NULL terminated (size smaller than its
+       length); copy and terminate part we are interested in. */
+    wchar_t* substr = PyMem_Malloc((size + 1) * sizeof(wchar_t));
+    if (substr == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    memcpy(substr, u, size * sizeof(wchar_t));
+    substr[size] = 0;
+
+    /* Convert given wide-character string to a character string */
+    size_t buffsize = wcstombs(NULL, substr, 0) + 1;
+    if (buffsize == (size_t)-1) {
+        PyMem_Free(substr);
+        PyErr_Format(PyExc_ValueError, "wcstombs() conversion failed");
+        return NULL;
+    }
+
+    /* Ensure we won't overflow the size. */
+    if (buffsize > (PY_SSIZE_T_MAX - 1)) {
+        PyMem_Free(substr);
+        PyErr_NoMemory();
+        return NULL;
+    }
+    char* buffer = PyMem_Malloc(buffsize * sizeof(char));
+    if (buffer == NULL) {
+        PyMem_Free(substr);
+        PyErr_NoMemory();
+        return NULL;
+    }
+
+    size_t res = wcstombs(buffer, substr, buffsize);
+    assert(res == buffsize - 1);
+
+    /* Convert character string to UTF32 encoded char32_t string.
+       Since wchar_t and char32_t have the same size on Solaris and one
+       wchar_t symbol corresponds to one UTF32 value, we can safely
+       reuse this buffer and skip additional allocation. */
+    char32_t* c32 = (char32_t*) substr;
+    mbstate_t state = {0};
+
+    Py_ssize_t i = 0;
+    char* ptr = buffer;
+    char* end = ptr + res + 1;
+
+    while (res = mbrtoc32(&(c32[i]), ptr, end - ptr, &state)) {
+        if (res == (size_t)-1 || res == (size_t)-2 || res == (size_t)-3) {
```

Comment:
I tried to find mbrtoc32() possible errors, but I only found -1. Would you mind to comment what -1, -2 and -3 means?

In fileutils.c, the -3 error can only come from Python, it means that the error handler is not supported by the codec.

Reply:
You can use DECODE_ERROR instead of -1 and INCOMPLETE_CHARACTER instead of -2.